### ISSUE TITLE
Hability to configure the type of the volume for the agent in the kubernetes backend

### ISSUE THREAD

**Issue by babykart** [woodpecker-ci/woodpecker#1594]
### Clear and concise description of the problem

Actually, woodpecker-agent creates a _PersistentVolumeClaim_ for the mount of the `/woodpecker` directory that is shared between the pods (steps) of the same pipeline.
This is occured even if you have `skip_clone: true` for example a pipeline that notices a chat channel.
In case of a cluster processing a lot of pipelines (every pipeline generate a _persistentVolumeClaim_ that will be destroyed at the end) these approach can overload the PersistentVolume backend of the Kubernetes cluster.

### Suggested solution

Drone CI uses a disk-based [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) for the `/drone` directory that is shared to all steps in a pipeline.
Convert the initial _PersistentVolumeClaim_ into a disk-based emptyDir.

### Alternative

An alernative would be the hability to configure the type of the `/woodpecker` directory :

```
WOODPECKER_BACKEND_K8S_STORAGE_TYPE=(PersistentVolume|EmptyDir|MemEmptyDir)
```

In this case the variable _WOODPECKER_BACKEND_K8S_VOLUME_SIZE_ would be usefull for all the types.

### Validations

- [X] Checked that the feature isn't part of the `next` version already [https://woodpecker-ci.org/faq#which-version-of-woodpecker-should-i-use]
- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by babykart** [comment]
Now that I better understand how the kubernetes backend works, this request is not applicable.
Indeed, woodpecker creates a pod for each step of the pipeline and therefore the PersistentVolume with the ReadWriteMany mode is mandatory in order to share at least the repository between the different pods which can be scheduled on different nodes in the cluster.
This behavior means that woodpecker-agent cannot be used in a cluster that does not have a persistent storage system.
Sorry I hadn't read the code, I thought the process was similar to drone, i.e., a pipeline is a pod and each step, clone included, is a container of that same pod so no problem using an emptyDir on disk or in memory since it is certain that each step will be executed on the same node where the pipeline was instantiated.
In short, if the implementation of the kubernetes backend is definitive, this feature can be rejected and it should be specified in the documentation that a persistent storage system is mandatory for the use of woodpecker with the kubernetes backend.
Sorry again for my misunderstanding.

**Comment by pi3ch** [comment]
I am facing issue of delay of up to 2m (depending on the volume driver) and steps take quite long to complete due to volume provisioning: https://github.com/woodpecker-ci/helm/issues/142 

> woodpecker creates a pod for each step of the pipeline and therefore the PersistentVolume with the ReadWriteMany mode is mandatory in order to share at least the repository between the different pods which can be scheduled on different nodes in the cluster.

From what I have gathered, woodpecker uses a single pod and each step is a container with in the same pod. So it should be possible to use EmptyDir similar to Drone. This makes pipeline to run a lot faster, and less costly.

**Comment by pat-s** [comment]
> From what I have gathered, woodpecker uses a single pod and each step is a container with in the same pod. So it should be possible to use EmptyDir similar to Drone. This makes pipeline to run a lot faster, and less costly.

No, each step has its own pod and the volume is required to share the repo/git data among these. In Drone CI I think (I don't run it anymore, can only recall from my memory), each pipeline has its own pod and uses different containers within that pod. So there, `emptyDir` works.

We can definitely think of changing the approach to a "one pod - multiple containers" one. This would allow us to get rid of the volume requirement (unless I am overlooking something which might cause other issues). 

@babykart @pi3ch If anyone of you want to contribute, I am happy to review a PR which changes the approach. Yet it might not be the easiest change. https://github.com/woodpecker-ci/woodpecker/blob/main/pipeline/backend/kubernetes/pod.go and friends are the source files of interest.

@pi3ch WRT to your specific issue, we first need to better understand what is causing it, as many others don't seem to face the long startup times (and I am not yet sure if it's really due to the volume mount). We can continue discussing in the other open issue.

**Comment by pi3ch** [comment]
> No, each step has its own pod and the volume is required to share the repo/git data among these. In Drone CI I think (I don't run it anymore, can only recall from my memory), each pipeline has its own pod and uses different containers within that pod. So there, `emptyDir` works.

Reason I said this was when I looked at event (to figure out the delay issue I had for RWX PVs), I only found one event where volume got attached to the pod. I didn't find any other volume attachment events. So I imagine this is because a pod running multiple containers for each step. I can do further testing.

**Comment by pat-s** [comment]
You can see the pods spawning for each step, they are individual pods. Each step has its own pod, this is also backed by the general implementation in the backend and the previous comments/findings from OP in this thread.

**Comment by zc-devs** [comment]
> This is occured even if you have skip_clone: true for example a pipeline that notices a chat channel

doesn't require changing approach to `one Pod per Workflow` as well as `ability to configure the type of the volume`.
Issue can be solved by disabling (not creating, not mounting) Volume if `skip_clone: true`.

**Comment by pat-s** [comment]
> This is occured even if you have skip_clone: true for example a pipeline that notices a chat channel.

> [Documentation](https://woodpecker-ci.org/docs/administration/backends/kubernetes). Should it be closed?

I don't understand the doc link here. Can you explain?
Users cannot decide to not mount the volume for a pipeline in the k8s backend/helm chart, currently.

**Comment by zc-devs** [comment]
> this request is not applicable

> this feature can be rejected and it should be specified in the documentation that a persistent storage system is mandatory for the use of woodpecker with the kubernetes backend

The first sentence of [documentation](https://woodpecker-ci.org/docs/administration/backends/kubernetes) says:

> The kubernetes backend executes steps inside standalone pods. A temporary PVC is created for the lifetime of the pipeline to transfer files between steps.

\+ the first part of the [comment](https://github.com/woodpecker-ci/woodpecker/issues/1594#issuecomment-1938174033)

=> this issue can be closed.

---

> Users cannot decide to not mount the volume for a pipeline in the k8s backend/helm chart, currently.

I believe. Not in Helm either. Should be there separate issue?
Or should this one be moved into `do not create Volume if skip_clone: true`?

**Comment by pat-s** [comment]
@zc-devs Thanks for clarifying!

I think there are different colours to this issue:

- Creating a volume if cloning is skipped is certainly unnecessary, yet the volume attachment happens in the helm chart and the chart isn't connected to listen to YAML inputs from WP core. So I don't think there can be a dynamic solution.
- The volume doesn't really matter if it is thrown away anyhow. I am not sure how many pods one has to create at the same time to overload the PV provisioner of the cluster. But this is actually a different topic.

The issue definitely has a point, i.e. if the volume type would be configurable and set to `emptyDir`, no PVC would be attached and if cloning is skipped, only the commands of the image of the step would be executed. This would happen without any local repo files then but these are not always needed anyhow. 

All of the above would anyhow happen if the "one-pod-many-containers" approach with a default `emptyDir` volume would be pursued. 

The overall proposal could be implemented but it would just then be another "quick fix" (similar to the proposal of RWX/RWO discussed in the other PR lately) compared to a full "one-pod-many-containers" option (which could live alongside the current "one pod per step" approach).

Overall the request is very similar to #3344 in the sense that OP wishes to use no volume at all to execute some commands while #3344 aimed for a RWO (which in the end is the same as an `emptyDir` if it's deleted anyhow after the pod is being removed.

There has been a new request in the helm chart repo today which also aims for removing the PV: https://github.com/woodpecker-ci/helm/pull/160. I am wondering a bit why so many requests targeting the same (but slightly different) come in within a few days 🤔

**Comment by zc-devs** [comment]
> yet the volume attachment happens in the helm chart and the chart isn't connected to listen to YAML inputs from WP core

> There has been a new request in the helm chart repo today which also aims for removing the PV

Sorry, I don't get nor the first, nor the last. I do not use Helm. And in my knowledge, I don't get how helm is related here at all.

---

Despite of anybody's wishes, Volume _is needed_ if you use `clone` step, point. 
That the volume is created, but not used, when `clone` is skipped is a flaw of current implementation. It may be legacy from the time, when we could not skip clone step (just a guess).
Even though it could be solved within new architecture, it should be tackled accordingly in current architecture without constructing spaceships, I think. At least, firstly.

---

This is all off-topic, considering the current Issues' title. Helm issues should be in Helm repo, new architecture discussion in #3366 and skip-clone-volume stuff in separate issue. I feel I'm repeating myself...

**Comment by pat-s** [comment]
> Sorry, I don't get nor the first, nor the last. I do not use Helm. And in my knowledge, I don't get how helm is related here at all.

I am sorry. You're right. Here it is about the pod created in `pod.go`. I thought and argued about the agent volume which is controlled by Helm.

> That the volume is created, but not used, when clone is skipped is a flaw of current implementation. It may be legacy from the time, when we could not skip clone step (just a guess).

That, or maybe, nobody has thought about skipping cloning at all to execute steps without repo files.

---

> This is all off-topic, considering the current Issues' title. Helm issues should be in Helm repo, new architecture discussion in https://github.com/woodpecker-ci/woodpecker/discussions/3366 and skip-clone-volume stuff in separate issue. I feel I'm repeating myself...

Agreed. Sorry for taking more time/words than initially hoped. Just a human ;)
